### PR TITLE
Add amazon project

### PR DESCRIPTION
**Pull request by RanHomri**
empty project skeleton
datascript was copied as is

### CODE REVIEW

File: examples/armbench-segmentation/tests/test_tensorleap.py
Comment (tomkoren21) [inline examples/armbench-segmentation/tests/test_tensorleap.py:1]: I generally like this flow less than the flow @daniellebenbashat  has used in her testing script:
https://github.com/tensorleap/tensorleap/blob/domain_gap/examples/domain_gap/tests/test_tensorleap.py

Testing should be explicit and call every metadata/metric/loss so the user can easily see how to test its own code IMO

Comment (yotamazriel) [reply examples/armbench-segmentation/tests/test_tensorleap.py:1]: We decided to do it like that now and add some comments to explain how to enrich it.
We'll have a better solution soon.

File: examples/armbench-segmentation/README.md
[the remark this answers is not in the document]
Comment (RanHomri) [reply examples/armbench-segmentation/README.md:1]: Added an edited version of @daniellebenbashat README

File: examples/armbench-segmentation/tensorleap.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    ann_file = os.path.join(DIR, IMG_FOLDER, "test.json")
+    fpath = _download(ann_file)
+    val = COCO(fpath)
+    x_val_raw = load_set(coco=val, load_union=LOAD_UNION_CATEGORIES_IMAGES)
+
+    train_size = min(len(x_train_raw), TRAIN_SIZE)
+    val_size = min(len(x_val_raw), VAL_SIZE)
+    np.random.seed(0)
+    train_idx, val_idx = np.random.choice(len(x_train_raw), train_size), np.random.choice(len(x_val_raw), val_size)
+    return [PreprocessResponse(length=train_size, data={'cocofile': train,
```
Comment (tip3x) [inline examples/armbench-segmentation/tensorleap.py:48]: I think it would be easier to understand like that:

`training_subset = PreprocessResponse(length=train_size, data={'cocofile': train, ...`
`validation_subset = PreprocessResponse(length=val_size, data={'cocofile': val, ...`
`return [training_subset, validation_subset]`

File: examples/armbench-segmentation/tensorleap.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+    mask_metric, over_segmented, under_segmented, metric_small_bb_in_under_segment, non_binary_over_segmented, \
+    non_binary_under_segmented, average_segments_num_over_segment, average_segments_num_under_segmented, \
+    over_segment_avg_confidence
+from armbench_segmentation.visualizers.visualizers import gt_bb_decoder, bb_decoder, \
+    under_segmented_bb_visualizer, over_segmented_bb_visualizer
+from armbench_segmentation.visualizers.visualizers_getters import mask_visualizer_gt, mask_visualizer_prediction
+
+
+# ----------------------------------------------------data processing--------------------------------------------------
+def subset_images():
```
Comment (tip3x) [inline examples/armbench-segmentation/tensorleap.py:32]:
```suggestion
def subset_images() -> List[PreprocessResponse]:
```

File: examples/armbench-segmentation/test_tensorleap.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import os
+from code_loader import LeapLoader
+
+
+def check_integration():
+    dir_path = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
```
Comment (tip3x) [inline examples/armbench-segmentation/test_tensorleap.py:6]:
```suggestion
    dir_path = str(Path(__file__).parent.parent)
```

File: examples/armbench-segmentation/armbench_segmentation/metrics.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    return tf.reduce_sum(loss_o, axis=0)[:, 0]  # shape of batch
+
+
+def mask_metric(bb_gt: tf.Tensor, detection_pred: tf.Tensor,
+                mask_gt: tf.Tensor, segmentation_pred: tf.Tensor):
+    _, _, _, loss_m = compute_losses(bb_gt, detection_pred, mask_gt, segmentation_pred)
+    return tf.reduce_sum(loss_m, axis=0)[:, 0]  # shape of batch
+
+
+def under_segmented(image, y_pred_bb, y_pred_mask, bb_gt, mask_gt):
```
Comment (tip3x) [inline examples/armbench-segmentation/armbench_segmentation/metrics.py:69]: typing

File: examples/armbench-segmentation/armbench_segmentation/metrics.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+
+def compute_losses(obj_true: tf.Tensor, od_pred: tf.Tensor,
+                   mask_true, instance_seg: tf.Tensor) -> Union[
+    Tuple[List[tf.Tensor], List[tf.Tensor], List[tf.Tensor]],
+    Tuple[List[tf.Tensor], List[tf.Tensor], List[tf.Tensor], List[tf.Tensor]]]:
+    """
+    Computes the sum of the classification (CE loss) and localization (regression) losses from all heads
+    """
+    res = CACHE_DICTS['loss'].get(str(obj_true) + str(od_pred) + str(mask_true) + str(instance_seg))
```
Comment (tip3x) [inline examples/armbench-segmentation/armbench_segmentation/metrics.py:21]: nice!

File: examples/armbench-segmentation/armbench_segmentation/__init__.py [hunk 6]
```diff
@@ -0,0 +1,4 @@
+CACHE_DICTS = {'loss': {},
```
Comment (tip3x) [inline examples/armbench-segmentation/armbench_segmentation/__init__.py:1]: let's try to reduce the memory of this cache (this is the reason for the OOM)

- use hashing instead of str(tensors)
- caching only the last batch for metrics that the visualizers don't use (for example the loss). also consider using the decorator `lru_cache(max_size=1)` for that
-  I think it would also be fine to cache the last sample for the visualizer metrics, we would get most of the speedup and reduce significantly the memory.

I think it can take 1-2 days, so maybe it's better to limit the number of samples in the script and do it when we have time.  @yotamazriel

**Review by tip3x — APPROVED** [review]
LGTM very nice! Added some linter comments.

We need to think about what to do with the cache memory issue
